### PR TITLE
etcd-dump-logs: add decoder support

### PR DESCRIPTION
**Pull request by wenjiaswe**
etcd-dump-logs: add decoder support

Added "stream-decoder" flag so user could set the decoder and decoder args they want to use to decode each WAL log entries. 

**Expected behavior of decoder:**
The decoder should be able to take hex encoded lines of binary as input (from etcd-dump-logs), process it, output the status and decoded data, preferably in the format of <DECODER_STATUS>|<DECODED_DATA>. (If the output is not in the preferable format, result will be put into decoded data column regardlessly)

**Expected behavior of etcd-dump-logs:**
When the stream-decoder is specified, etcd-dump-logs will feed the hex encoded lines of WAL LOG binary into the decoder and output the decoder_status and decoded_data for each entry.
 
**Two simple decoders were added for testing and demonstrating purpose:**
Both decoders ([decoder_correctoutputformat.sh](https://github.com/wenjiaswe/etcd/blob/decoder-support/tools/etcd-dump-logs/testdecoder/decoder_correctoutputformat.sh) and [decoder_wrongoutputformat.sh](https://github.com/wenjiaswe/etcd/blob/decoder-support/tools/etcd-dump-logs/testdecoder/decoder_wrongoutputformat.sh) ) are simply converting 1234567890 to abcdefghij. decoder_correctoutputformat.sh outputs decoder_status and decoded_data in the preferable format as above (<DECODER_STATUS>|<DECODED_DATA>), while decoder_wrongoutputformat.sh does not. The different result from etcd-dump-logs can be found in [decoder_correctoutputformat.output](https://github.com/wenjiaswe/etcd/blob/decoder-support/tools/etcd-dump-logs/expectedoutput/decoder_correctoutputformat.output) and [decoder_wrongoutputformat.output](https://github.com/wenjiaswe/etcd/blob/decoder-support/tools/etcd-dump-logs/expectedoutput/decoder_wrongoutputformat.output).

**Combination usage of etcd-dump-logs and Auger:**
Other than the two "fake" decoder above, I would like to mention a real usage of this added feature using Auger as decoder. [Auger](https://github.com/jpbetz/auger) (please refer to Auger link for further details if you are interested) is a very useful tool in kubernetes community to inspect and analyze kubernetes objects stored in etcd. Now with this change, we could use Auger as the decoder to decode kubernetes objects in WAL log and have them output nicely. Here is the example of using etcd-dump-logs and Auger.

> $ etcd-dump-logs -stream-decoder "auger decode --output json --batch-process true"  datadir
> Snapshot:
> empty
> Start dupmping log entries from snapshot.
> WAL metadata:
> nodeID=2f7fbfd36802ff7a clusterID=bea638dce5096eef term=2 commitIndex=3358 vote=2f7fbfd36802ff7a
> WAL entries:
> lastIndex=3358
> term	     index	type	data	decoder_status	decoded_data
>    1	         1	conf	method=ConfChangeAddNode id=2f7fbfd36802ff7a		AUGER_ERROR:error decoding from application/json: Object 'Kind' is missing in '{"id":3422665156528897914,"peerURLs":["https://kubernetes-master:2380"],"name":"etcd-kubernetes-master"}'
>    2	         2	norm		AUGER_ERROR:error reading input, does not appear to contain valid JSON or binary data	
>    2	         3	norm	method=PUT path="/0/members/2f7fbfd36802ff7a/attributes" val="{\"name\":\"etcd-kubernetes-master\",\"clientURLs\":[\"http://127.0.0.1:2379\"]}"	AUGER_ERROR:error reading input, does not appear to contain valid JSON or binary data	
>    2	         4	norm	method=PUT path="/0/version" val="3.2.0"	AUGER_ERROR:error reading input, does not appear to contain valid JSON or binary data	
>    2	         5	norm	method=QGET path=""	AUGER_ERROR:error reading input, does not appear to contain valid JSON or binary data	
>    2	         6	norm	method=QGET path=""	AUGER_ERROR:error reading input, does not appear to contain valid JSON or binary data	
>    2	         7	norm	header:<ID:18409135401817889912 > txn:<compare:<target:MOD key:"/registry/ranges/serviceips" mod_revision:0 > success:<request_put:<key:"/registry/ranges/serviceips" value:"k8s\000\n\025\n\002v1\022\017RangeAllocation\022#\n\022\n\000\022\000\032\000\"\000*\0002\0008\000B\000z\000\022\01310.0.0.0/16\032\000\032\000\"\000" > > failure:<request_range:<key:"/registry/ranges/serviceips" > > > 	AUGER_OK	{"kind":"RangeAllocation","apiVersion":"v1","metadata":{"creationTimestamp":null},"range":"10.0.0.0/16","data":""}
>    2	         8	norm	header:<ID:18409135401817889914 > txn:<compare:<target:MOD key:"/registry/ranges/servicenodeports" mod_revision:0 > success:<request_put:<key:"/registry/ranges/servicenodeports" value:"k8s\000\n\025\n\002v1\022\017RangeAllocation\022#\n\022\n\000\022\000\032\000\"\000*\0002\0008\000B\000z\000\022\01330000-32767\032\000\032\000\"\000" > > failure:<request_range:<key:"/registry/ranges/servicenodeports" > > > 	AUGER_OK	{"kind":"RangeAllocation","apiVersion":"v1","metadata":{"creationTimestamp":null},"range":"30000-32767","data":""}

**README.md is added.**
As etcd-dump-logs is getting more powerful and complicated, I think it's easier to have a README.md to explain the usage.

@gyuho @xiang90 @jpbetz

### CODE REVIEW

File: tools/etcd-dump-logs/README.md [hunk 1]
```diff
@@ -0,0 +1,141 @@
...
+	    and output a hex encoded line of binary for each input line
+
+```
+
+#### etcd-dump-logs -entry-type <ENTRY_TYPE_NAME(S)> [data dir]
+
+Lists all the interested entries from WAL log.
+
+```
+
```
Comment (gyuho) [inline tools/etcd-dump-logs/README.md]: remove this unneeded blank line at 39?

File: tools/etcd-dump-logs/README.md [hunk 2]
```diff
@@ -0,0 +1,141 @@
...
+lastIndex=34
+term	     index	type	data
+  25	        31	norm	ID:26 auth_role_get:<role:"role3" >
+  26	        32	norm	ID:27 auth_role_grant_permission:<name:"role3" perm:<permType:WRITE key:"Keys" range_end:"RangeEnd" > >
+  27	        33	norm	ID:28 auth_role_revoke_permission:<role:"role3" key:"key" range_end:"rangeend" >
+  27	        34	norm	???
+Entry types () count is : 4
+```
+
+[decoder_correctoutputformat.sh]: https://github.com/wenjiaswe/etcd/blob/decoder-support/tools/etcd-dump-logs/testdecoder/decoder_correctoutputformat.sh
```
Comment (gyuho) [inline tools/etcd-dump-logs/README.md]: `[decoder_correctoutputformat.sh]: ./testdecoder/decoder_correctoutputformat.sh`?

File: tools/etcd-dump-logs/etcd-dump-log_test.go [hunk 3]
```diff
@@ -45,6 +45,9 @@ func TestEtcdDumpLogEntryType(t *testing.T) {
 		t.Skipf("%q does not exist", dumpLogsBinary)
 	}
 
+	decoder_correctoutputformat := path.Join(binDir + "/testdecoder/decoder_correctoutputformat.sh")
```
Comment (gyuho) [inline tools/etcd-dump-logs/etcd-dump-log_test.go]: `filepath.Join(binDir, "/testdecoder/decoder_correctoutputformat.sh")`?

File: tools/etcd-dump-logs/etcd-dump-log_test.go [hunk 4]
```diff
@@ -45,6 +45,9 @@ func TestEtcdDumpLogEntryType(t *testing.T) {
...
+	decoder_wrongoutputformat := path.Join(binDir + "/testdecoder/decoder_wrongoutputformat.sh")
```
Comment (gyuho) [inline tools/etcd-dump-logs/etcd-dump-log_test.go]: `filepath.Join(binDir, "/testdecoder/decoder_wrongoutputformat.sh")`?

File: tools/etcd-dump-logs/main.go [hunk 5]
```diff
@@ -31,6 +35,8 @@ import (
 	"github.com/coreos/etcd/wal/walpb"
 
 	"go.uber.org/zap"
+	"bytes"
```
Comment (gyuho) [inline tools/etcd-dump-logs/main.go]: Can you move `"bytes"` and `"os"` at the same level as `"encoding/hex"`? Just minor `gofmt` issue.

File: tools/etcd-dump-logs/README.md [hunk 6]
```diff
@@ -0,0 +1,141 @@
+### etcd-dump-logs
+
+etcd-dump-logs inspects etcd db files.
+
+```
+
```
Comment (gyuho) [inline tools/etcd-dump-logs/README.md]: Remove unnecessary blank line at 6?

File: tools/etcd-dump-logs/main.go [hunk 7]
```diff
@@ -301,7 +336,53 @@ func listEntriesType(entrytype string, ents []raftpb.Entry) {
...
 		}
 	}
+
+	stdin.Close()
+	err = cmd.Wait()
+	if streamdecoder != "" {
+		if err != nil {
+			log.Panic(err)
+		}
+		os.Stderr.WriteString("decoder stderr: " + stderr.String())
```
Comment (gyuho) [inline tools/etcd-dump-logs/main.go]:
```go
if stderr.String() != "" {
  os.Stderr.WriteString("decoder stderr: " + stderr.String())
}
```

?

Comment (wenjiaswe) [reply tools/etcd-dump-logs/main.go]: Thanks! Fixed!

File: tools/etcd-dump-logs/README.md [hunk 8]
```diff
@@ -0,0 +1,131 @@
...
+   3	         2	norm	noop	decoder output format is not right, print output anyway	jhjjabjjaajfbfgjfagdfhcjbbahgbbbfhfegibbcabbfhffbbbcbbfhfibbcaebbbgiffbbedgdbhjacbjjchjjdjjjdhjiejjjehjafjjjfhjjgjjjghjahjjajjhhjajj
+   3	         3	norm	method=QGET path="/path1"	decoder output format is not right, print output anyway	jhjaabjdeadgdeedaajfbfgjfagdfhcabbacgbbbcjbbcabbcabbbcbbcbbbcaebbbccbbedgdbhjjcbjjchjjdjjjdhjiejjjehjafjjjfhjjgjjjghjahjjajjhhjajj
+   7	         4	norm	ID:8 txn:<success:<request_delete_range:<key:"a" range_end:"b" > > failure:<request_delete_range:<key:"a" range_end:"b" > > > 	decoder output format is not right, print output anyway	jhjhcbadabjhaajfjajafaabjafbaajhaajfjajafaabjafb
+   8	         5	norm	ID:9 compaction:<physical:true > 	decoder output format is not right, print output anyway	jhjicajbajja
+   9	         6	norm	ID:10 lease_grant:<TTL:1 ID:1 > 	decoder output format is not right, print output anyway	jhjadbjdjhjaajja
+  12	         7	norm	ID:13 auth_enable:<> 	decoder output format is not right, print output anyway	jhjdcbcejj
+  27	         8	norm	???	decoder output format is not right, print output anyway	cf
+decoder stderr:
```
Comment (gyuho) [inline tools/etcd-dump-logs/README.md]: Remove this?

File: tools/etcd-dump-logs/README.md [hunk 9]
```diff
@@ -0,0 +1,131 @@
...
+term	     index	type	data	decoder_status	decoded_data
+   1	         1	conf	method=ConfChangeAddNode id=2	ERROR	jhjaajjjahjbbbjj
+   3	         2	norm	noop	OK	jhjjabjjaajfbfgjfagdfhcjbbahgbbbfhfegibbcabbfhffbbbcbbfhfibbcaebbbgiffbbedgdbhjacbjjchjjdjjjdhjiejjjehjafjjjfhjjgjjjghjahjjajjhhjajj
+   3	         3	norm	method=QGET path="/path1"	OK	jhjaabjdeadgdeedaajfbfgjfagdfhcabbacgbbbcjbbcabbcabbbcbbcbbbcaebbbccbbedgdbhjjcbjjchjjdjjjdhjiejjjehjafjjjfhjjgjjjghjahjjajjhhjajj
+   7	         4	norm	ID:8 txn:<success:<request_delete_range:<key:"a" range_end:"b" > > failure:<request_delete_range:<key:"a" range_end:"b" > > > 	OK	jhjhcbadabjhaajfjajafaabjafbaajhaajfjajafaabjafb
+   8	         5	norm	ID:9 compaction:<physical:true > 	ERROR	jhjicajbajja
+   9	         6	norm	ID:10 lease_grant:<TTL:1 ID:1 > 	ERROR	jhjadbjdjhjaajja
+  12	         7	norm	ID:13 auth_enable:<> 	ERROR	jhjdcbcejj
+  27	         8	norm	???	ERROR	cf
+decoder stderr:
```
Comment (gyuho) [inline tools/etcd-dump-logs/README.md]: Remove this?

**Comment by gyuho** [comment]
Manually addressed README.md typos. Merging. Thanks.

**Comment by wenjiaswe** [comment]
Thanks, @gyuho !

**Pull request merged**